### PR TITLE
[PowerPC][GlobalMerge] Reduce TOC usage by merging internal and private global data

### PR DESCRIPTION
**Pull request by amy-kwan**
This patch aims to reduce TOC usage by merging internal and private global data.

Moreover, we also add the GlobalMerge pass within the PPCTargetMachine pipeline, which is disabled by default. This transformation can be enabled by -ppc-global-merge.

### CODE REVIEW

File: llvm/include/llvm/CodeGen/GlobalMerge.h [hunk 1]
```diff
@@ -28,6 +28,10 @@ struct GlobalMergeOptions {
   bool MergeConst = false;
   /// Whether we should merge global variables that have external linkage.
   bool MergeExternal = true;
+  /// Whether we should merge global variables that have private linkage.
+  bool MergePrivateGlobals = false;
```
Comment (efriedma-quic) [inline llvm/include/llvm/CodeGen/GlobalMerge.h:32]: MergePrivateGlobals is unused?

File: llvm/lib/CodeGen/GlobalMerge.cpp [hunk 2]
```diff
@@ -720,8 +731,14 @@ bool GlobalMergeImpl::run(Module &M) {
...
-                                  bool MergeExternalByDefault) {
+                                  bool MergeExternalByDefault,
+                                  bool MergePrivateByDefault,
+                                  bool MergeConstantByDefault) {
   bool MergeExternal = (EnableGlobalMergeOnExternal == cl::BOU_UNSET) ?
     MergeExternalByDefault : (EnableGlobalMergeOnExternal == cl::BOU_TRUE);
-  return new GlobalMerge(TM, Offset, OnlyOptimizeForSize, MergeExternal);
+  bool MergeConstant = EnableGlobalMergeOnConst.getNumOccurrences() > 0
+                           ? EnableGlobalMergeOnConst
+                           : MergeConstantByDefault;
```
Comment (redstar) [inline llvm/lib/CodeGen/GlobalMerge.cpp:741]:
```suggestion
  bool MergeConstant = EnableGlobalMergeOnConst || MergeConstantByDefault;
```

File: llvm/include/llvm/CodeGen/GlobalMerge.h (hunk 1 above)
Comment (chenzheng1030) [reply llvm/include/llvm/CodeGen/GlobalMerge.h:32]: +1

Like the internal global variables, private global variables should always be merged and should not be guarded under an option. You've already done this in https://github.com/llvm/llvm-project/pull/101222 :)

**Review by redstar — APPROVED** [review]
LGTM.

**Review by chenzheng1030 — APPROVED** [review]
LGTM. Thanks very much for enabling this pass on PPC.

**Pull request merged**